for #5271: store loaded tabs when finished loading

MatthewTighe opened the pull request:
### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

jonalmeida commented on components/feature/syncedtabs/src/test/java/mozilla/components/feature/syncedtabs/storage/SyncedTabsStorageTest.kt:
```diff
@@ -229,4 +235,69 @@ class SyncedTabsStorageTest {
+        feature.start()
+
+        // Change a tab besides loading it
+        store.dispatch(ContentAction.UpdateProgressAction("tab1", 50)).joinBlocking()
```
Shouldn't we verify that no other `store` calls were made after the init? The state changes from `UpdateProgressAction` are not observed from the `SyncedTabsStorage`, so we should assert that this change does not call `store` again.

MatthewTighe replied:
I thought verifying that the `store` call occurred exactly once would be good enough to indicate that. Is there something else you'd suggest? `Mockito.any` has a Kotlin compatibility bug, so I'm not able to do the following:
```
        verify(tabsStorage, times(1)).store(
            eq(listOf(
                Tab(history = listOf(TabEntry(title = "", url = "https://www.mozilla.org", iconUrl = null)), active = 0, lastUsed = 123L),
                Tab(history = listOf(TabEntry(title = "", url = "https://www.foo.bar", iconUrl = null)), active = 0, lastUsed = 124L)
            ))
        )
        verify(tabsStorage, never()).store(any())
```

I tried with the custom `any()` matcher defined in the project as well, but it doesn't seem to allow sequenced verification like Mockito normally does.

jonalmeida replied:
I think what you're looking for in this test is `reset`; it isn't a recommended approach for regular testing, but it helps for this one:

```kotlin

        // Tabs are only stored when initial state is collected, since they are already loaded
        verify(tabsStorage, times(1)).store(
            listOf(
                Tab(history = listOf(TabEntry(title = "", url = "https://www.mozilla.org", iconUrl = null)), active = 0, lastUsed = 123L),
                Tab(history = listOf(TabEntry(title = "", url = "https://www.foo.bar", iconUrl = null)), active = 0, lastUsed = 124L)
            )
        )

        // reset the mock so there are no verifications or prior calls on the mock.
        reset(tabsStorage)

        // Change a tab besides loading it
        store.dispatch(ContentAction.UpdateProgressAction("tab1", 50)).joinBlocking()

        verify(tabsStorage, never()).store(any())
```

MatthewTighe replied:
TY for the added detail! That does seem to work

jonalmeida commented on components/feature/syncedtabs/src/test/java/mozilla/components/feature/syncedtabs/storage/SyncedTabsStorageTest.kt:
```diff
@@ -75,7 +81,7 @@ class SyncedTabsStorageTest {
         store.dispatch(TabListAction.RemoveAllPrivateTabsAction).joinBlocking()
 
-        verify(tabsStorage).store(
+        verify(tabsStorage, atLeastOnce()).store(
```
Avoid using an unbound verification because future changes to the flow could lead to more unexpected calls and we could catch them here, if so.

We know that we should expect 2 calls - initial and updated states. So let's verify that.

```suggestion
        verify(tabsStorage, times(2)).store(
```

MatthewTighe replied:
Good point! I updated this here and below

jonalmeida commented on components/feature/syncedtabs/src/test/java/mozilla/components/feature/syncedtabs/storage/SyncedTabsStorageTest.kt:
```diff
@@ -95,7 +101,7 @@ class SyncedTabsStorageTest {
         store.dispatch(TabListAction.RemoveAllPrivateTabsAction).joinBlocking()
 
-        verify(tabsStorage).store(
+        verify(tabsStorage, atLeastOnce()).store(
```
Same comment as above.

```suggestion
        verify(tabsStorage, times(2)).store(
```

jonalmeida reviewed (approved):
Minor updates needed for the tests, but looks good!

MatthewTighe commented:
@jonalmeida I had a question about one of your comments above, but I otherwise addressed the feedback. Also added a commit that implements syncing when tabs are stored. Would it be better to submit that commit in a separate PR?

I considered a couple different approaches here for syncing:
- Using a middleware. This approach felt error-prone, since an exhaustive list of actions is larger than just observing state changes, especially as additional actions are introduced. Using a middleware also doesn't make guarantees about race conditions.
- Using a `Store` `Observer`. This felt like a better fit, but ultimately has the same race condition issue.

Adding the `syncNow` call to the `storage` feels like it may be injecting a bit too much responsibility into the class, but it allows for easy guarantees of ordering. Since we are `debounce`-ing the `store` calls along with the `syncNow` calls, we'll also be avoiding some number of disk writes.

jonalmeida commented on components/feature/syncedtabs/src/main/java/mozilla/components/feature/syncedtabs/storage/SyncedTabsStorage.kt:
```diff
@@ -27,26 +33,34 @@ import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifChanged
-                tabsStorage.store(tabs)
             }
+                .debounce(debounceMillis)
+                .collect { tabs ->
+                    tabsStorage.store(tabs)
+                    accountManager.syncNow(
+                        reason = SyncReason.User,
+                        customEngineSubset = listOf(SyncEngine.Tabs)
+                    )
+                }
```
Having there does mean this is no longer a "storage" later alone any more and it does other things, which you've identified already.

How about we try and create a delegate that would perform post-collect tasks. We have a SyncTabsController (update it to with a syncNow `customEngineSubset`), so we could sort of use that here too.

```kotlin
class SyncedTabsStorage(
    private val accountManager: FxaAccountManager,
    private val store: BrowserStore,
    private val tabsStorage: RemoteTabsStorage = RemoteTabsStorage(),
    // This could be an interface too that calls a controller.syncNow() instead of an anonymous lambda.
    private val onStoreComplete: () -> Unit = { accountManager.syncNow() } 
)
```

```suggestion
                .collect { tabs ->
                    tabsStorage.store(tabs)
                    onStoreComplete()
                }
```

MatthewTighe replied:
Just to clarify, are you suggesting we inject the `SyncTabsController` into the storage and use that for the default `onStoreComplete` delegate, or just to use the suggested code change above?

MatthewTighe replied:
I'll add the custom sync engine subset to the `SyncTabsController` implementation as well

jonalmeida commented on components/feature/syncedtabs/src/main/java/mozilla/components/feature/syncedtabs/storage/SyncedTabsStorage.kt:
```diff
@@ -27,26 +33,34 @@ import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifChanged
-            }.collect { tabs ->
-                tabsStorage.store(tabs)
             }
+                .debounce(debounceMillis)
```
(not sure if I missed this line the first pass)

With the account manager's debouncing mechanism that already exists, I'm not sure there is more added value here any more. Saving on disk writes means potentially losing state that we could be syncing the next time the device is able to.

MatthewTighe replied:
The account manager's debouncing mechanism seems to operate in the opposite manner to the `Flow.debounce` operator. That is:

- flow debouncing emits the latest value based on the timeout
- sync manager will ignore additional sync requests if they are within a debounce period

I think we want the behavior of the of flow debouncing, regardless of whether we want to use this specific implementation. Tab changes can happen rapidly due to different actions, so we want to wait for a "final" state before initiating a sync. 

If we aren't concerned about disk write frequency, I can rework the debounce mechanism to operate specifically on the sync requests

jonalmeida replied:
> I think we want the behavior of the of flow debouncing, regardless of whether we want to use this specific implementation. Tab changes can happen rapidly due to different actions, so we want to wait for a "final" state before initiating a sync.

Wouldn't the sync manager's debounce support this behaviour too? We get rapid tab changes, the sync manager ignores most requests until the debounce times out and the next sync is allowed. Is that the right understanding? 

>If we aren't concerned about disk write frequency, I can rework the debounce mechanism to operate specifically on the sync requests

I think we only need to remove the flow `debounce`. 🤔

MatthewTighe replied:
One thing to consider is that the sync manager debounce timer is also 5 minutes, and isn't currently configurable. My understanding of the sync manager debounce could lead to this situation:

- storage observes initial state, calls `syncNow(debounce = true)`
- unloaded tabs load
- selected tab is changed (any number of times within 5 minutes)

We would debounce every change that happened _after_ the initial state load until the 5 minute window had passed. I think we want the opposite: to ignore every change during a window until the last one, so we sync the most recent data.

jonalmeida replied:
Thanks for the explanation and doing the homework to study the differences between the two. I think your analysis is correct and we're better off with both these debouncers.

jonalmeida reviewed (commented):
Nice, I think we're almost there. I left some considerations and we can try to keep the implementation simple at the same time too.

MatthewTighe commented on components/feature/syncedtabs/src/main/java/mozilla/components/feature/syncedtabs/storage/SyncedTabsStorage.kt:
```diff
@@ -83,4 +97,16 @@ class SyncedTabsStorage(
+        val loadedTabs: List<TabSessionState>,
+    )
+
+    private fun BrowserState.toSyncComponents() = SyncComponents(
```
I like `SyncTabState` more for sure, and don't think I have a better suggestion. I was trying to communicate "these are the components that are required for updates", and couldn't come up with a better succinct description.

MatthewTighe commented:
Made a few more updates based on feedback. I'm not sure the `debounce` conversation is totally resolved yet, but let me know if the decision is to just remove it and rely on the `SyncManager`'s debounce mechanism.

jonalmeida commented on components/feature/syncedtabs/src/main/java/mozilla/components/feature/syncedtabs/storage/SyncedTabsStorage.kt:
```diff
@@ -23,30 +27,42 @@ import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifChanged
+    private val tabsStorage: RemoteTabsStorage = RemoteTabsStorage(),
+    private val debounceMillis: Long = 1000L,
+    private val onStoreComplete: () -> Unit = {}
 ) : SyncedTabsProvider {
     private var scope: CoroutineScope? = null
 
     /**
      * Start listening to browser store changes.
      */
+    @OptIn(FlowPreview::class)
     fun start() {
         scope = store.flowScoped { flow ->
-            flow.ifChanged { it.tabs }.map { state ->
+            flow.ifChanged { it.toSyncTabState() }.map { state ->
                 // TO-DO: https://github.com/mozilla-mobile/android-components/issues/5179
                 val iconUrl = null
-                state.tabs.filter { !it.content.private }.map { tab ->
+                state.tabs.filter { !it.content.private && !it.content.loading }.map { tab ->
                     // TO-DO: https://github.com/mozilla-mobile/android-components/issues/1340
                     val history = listOf(TabEntry(tab.content.title, tab.content.url, iconUrl))
                     Tab(history, 0, tab.lastAccess)
                 }
-            }.collect { tabs ->
-                tabsStorage.store(tabs)
             }
+                .debounce(debounceMillis)
```
super minor nit only if you're updating your PR for other reasons:

I didn't realize the `map` was on the same line (and the linter doesn't complain) so the braces are super weird and I can't un-see it. 🙈 

Could you move it to the next line and reformat that block to be tabbed up?

```kotlin
            flow.ifChanged { it.toSyncTabState() }
                .map { state ->
                    // TO-DO: https://github.com/mozilla-mobile/android-components/issues/5179
                    val iconUrl = null
                    state.tabs.filter { !it.content.private && !it.content.loading }.map { tab ->
                        // TO-DO: https://github.com/mozilla-mobile/android-components/issues/1340
                        val history = listOf(TabEntry(tab.content.title, tab.content.url, iconUrl))
                        Tab(history, 0, tab.lastAccess)
                    }
                }
                .debounce(debounceMillis)

```

MatthewTighe replied:
Oh yeah, that was super gross. Thanks for calling it out!

jonalmeida commented on components/feature/syncedtabs/src/main/java/mozilla/components/feature/syncedtabs/controller/DefaultController.kt:
```diff
@@ -58,7 +59,7 @@ internal class DefaultController(
         scope.launch {
             accountManager.withConstellation { refreshDevices() }
-            accountManager.syncNow(SyncReason.User)
+            accountManager.syncNow(SyncReason.User, customEngineSubset = listOf(SyncEngine.Tabs))
```
Thanks for doing this too!

jonalmeida reviewed (approved):
Thanks for being patient, this looks good! :shipit:
